Add xiaomi_miio humidifier entity and sensors

jbouwh opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Xiaomi miio Humidifiers will no longer be presented as a fan entity and be replaced with a humidifier entity and sensors.
Uses need to update their automations since entity_id's will change.
Humidifier specific services for setting `child_lock`, `buzzer`, `motor_speed`, `dry_mode` and `led_brightness` are not available any more. 

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR intents to transform the Xiaomi humidifier fan entities to humidity entities with additional sensors for water level (if supported, temperature and humidity. In related PR https://github.com/home-assistant/core/pull/51791 the existing fan entities got percentage based speeds and preset modes. But that will not be needed when using the humidifier integration.

New features are:
- introducing the humidifier entity type which enables setting the target humidity.
- introducing sensors for `humidity`, `temperature` and `water_level`*

As a follow up to this PR switches or other integrations to enable `dry`, `child_lock`, `buzzer` or `led_brightness` could be added.

*If supported

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io


> `MODELS_ALL_DEVICES`

Added `MODELS_HUMIDIFIER` to `MODELS_ALL_DEVICES`

jbouwh committed:
Humidifier and attr as humidifier and sensors

jbouwh committed:
Humidifier and attr as humidifier and sensors

jbouwh committed:
Set mode to auto after setting target_humidity

jbouwh committed:
Set mode to auto after setting target_humidity

jbouwh committed:
Only return target_humidity in MODE_AUTO

jbouwh committed:
Only return target_humidity in MODE_AUTO

jbouwh committed:
Implement humidity steps

jbouwh committed:
Implement humidity steps

bieniu commented:
You overtook me, I spent yesterday evening working on the same change ;)

bieniu commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,490 @@
+                name, air_humidifier, config_entry, unique_id
+            )
+        elif model.startswith("zhimi.humidifier."):
+            air_humidifier = AirHumidifier(host, token, model=model)
```
You create `miio` device object for `humidifier` entity and for each sensor `entity` and for `fan` entity. All these objects poll the device.
IMO one `miio` device object should be created for all platforms.

jbouwh replied:
You are right, I will have a look.

jbouwh replied:
I have now moved creating the miio device to __init__.py where it is cached to be used during the platforms setup.

bieniu commented:
IMO `fan` platform should be removed for humidifiers and this PR should be marked as a breaking change.

jbouwh commented:
> 
> 
> IMO `fan` platform should be removed for humidifiers and this PR should be marked as a breaking change.

The first commits were actually leaving away fan. I agree this should become a breaking change. I have not touched `fan.py`, but I could omit 'fan' from the humidifier platform setup. The reason to leave it for now is not to interfere with my other PR which implements percentage based speeds and preset_modes. Afterwards we can remove the humidifiers from the fan platform.

jbouwh committed:
Do not load fan platform for humidifiers

jbouwh committed:
Do not load fan platform for humidifiers

jbouwh committed:
Typo

jbouwh committed:
Typo

jbouwh committed:
Reuse miio_device

jbouwh committed:
Reuse miio_device

starkillerOG commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,491 @@
+CONF_MODEL = "model"
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Do not add yaml import to a new platform

starkillerOG commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,491 @@
+AIRHUMIDIFIER_SERVICE_SCHEMA = vol.Schema({vol.Optional(ATTR_ENTITY_ID): cv.entity_ids})
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
```
This yaml import can be removed

starkillerOG commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -97,6 +112,38 @@ class SensorType:
     ),
 }
 
+HUMIDIFIER_SENSOR_TYPES = {
```
I think I would just make one SENSOR_TYPES dict for gateway and humidifier and all other devices, that will make it a lot easier in the future.

starkillerOG commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -117,6 +164,9 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up the Xiaomi sensor from a config entry."""
     entities = []
 
+    model = config_entry.data[CONF_MODEL]
```
Move this inside the if == CONF_DEVICE

starkillerOG commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +195,127 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+
+            return False
+
+    async def async_update(self):
```
We schould actually use a DataUpdateCoordinator, that prevents duplicate API calls.

starkillerOG reviewed (commented):
Nice work!
Some first feedback from my side.

jbouwh commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -30,10 +33,14 @@
 AIR_MONITOR_PLATFORMS = ["air_quality", "sensor"]
 
+# global MIIO_DEVICE_CACHE
+MIIO_DEVICE_CACHE = {}
```
I think this will work. I have local code running, without data update coordinator. Will do that part later

jbouwh committed:
not using global variables

jbouwh committed:
not using global variables

jbouwh committed:
remove yaml import

jbouwh committed:
remove yaml import

rytilahti commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,462 @@
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return DEVICE_CLASS_HUMIDIFIER
```
Prefer entity class attributes over properties for static values: https://developers.home-assistant.io/docs/core/entity#entity-class-attributes

rytilahti commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,462 @@
+        return SUPPORT_MODES
+
+    @property
+    def should_poll(self):
+        """Poll the device."""
+        return True
```
Is this not a default setting, i.e., no need for having it at all?

rytilahti commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +195,129 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+
+        return value
+
+    async def _try_command(self, mask_error, func, *args, **kwargs):
+        """Call a miio device command handling error messages."""
+        try:
+            result = await self.hass.async_add_executor_job(
+                partial(func, *args, **kwargs)
+            )
+
+            _LOGGER.debug("Response received from miio device: %s", result)
+
+            return result == SUCCESS
+        except DeviceException as exc:
+            if self._available:
+                _LOGGER.error(mask_error, exc)
+                self._available = False
+
+            return False
```
This looks like copy&paste from above. I think this is shared among all entities(?), so this should rather belong to `XiaomiMiioEntity` to avoid code duplication.

rytilahti commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +195,129 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
     async_add_entities(entities, update_before_add=True)
 
 
+class XiaomiHumidifierSensor(XiaomiMiioEntity, SensorEntity):
```
Is there something so special for humidifer sensors versus regular sensors, that this should not be a `XiaomiSensor` and re-used by other platforms?

rytilahti commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,462 @@
+            self._state = False
+            self._skip_update = True
+
+    async def async_translate_humidity(self, humidity):
```
As this does no I/O (right?), it doesn't need to be defined as coroutine.

rytilahti reviewed (commented):
Thanks for the PR! I added some comments here and there.

jbouwh committed:
Bring DataUpdateCoordinator to entity classes

jbouwh committed:
Bring DataUpdateCoordinator to entity classes

jbouwh committed:
Remove ATTR_MIO_DEVICE const

jbouwh committed:
Remove ATTR_MIO_DEVICE const

jbouwh committed:
create coordinator only for humidifier models

jbouwh committed:
create coordinator only for humidifier models

jbouwh committed:
Remove yaml import

jbouwh committed:
Remove yaml import

jbouwh committed:
attributes over properties

jbouwh committed:
attributes over properties

jbouwh committed:
move _try_command to XiaomiMiioEntity class

jbouwh committed:
move _try_command to XiaomiMiioEntity class

jbouwh committed:
translate_humidity not a coroutine

jbouwh committed:
translate_humidity not a coroutine

jbouwh committed:
use XiaomiGenericSensor no special class

jbouwh committed:
use XiaomiGenericSensor no special class

jbouwh reviewed (commented):
All comments were processed for now.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+    """Set up the Fan from a config entry."""
+    entities = []
+
+    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
```
Invert and return early. And do this check first before anything else.

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+                air_humidifier,
+                config_entry,
+                unique_id,
+                hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
```
Store the coordinator in a local variable first outside the checks.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+                hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
+            )
+
+        hass.data[DATA_KEY][host] = entity
```
Don't store the entity.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+
+    _attr_device_class = DEVICE_CLASS_HUMIDIFIER
+    _attr_supported_features = SUPPORT_MODES
+    _attr_should_poll = True
```
We shouldn't poll. The point with the coordinator is that it handles the polling.

MartinHjelmare replied:
```suggestion
```

jbouwh replied:
Done

jbouwh replied:
Somehow it did not remove it at all. Removed the line now.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+        self.coordinator.async_add_listener(self.update)
+
+    @callback
+    def update(self):
```
It's best to use a different name. `update` has special features in some entities.
```suggestion
    def _update(self):
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+        self._state_attrs.update(
+            {attribute: None for attribute in self._available_attributes}
+        )
+        self.coordinator.async_add_listener(self.update)
```
```suggestion
        self.coordinator.async_add_listener(self._update)
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+    def mode(self):
+        """Return the current mode."""
+        if self._state:
+            return AirhumidifierOperationMode(self._state_attrs[ATTR_MODE]).name
```
Please always return a value specifically when returning a value somewhere.
```suggestion
            return AirhumidifierOperationMode(self._state_attrs[ATTR_MODE]).name
        return None
```

jbouwh replied:
Updated this one and on some more places with the same usecase

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+                else None
+            )
+
+    async def async_set_humidity(self, humidity) -> None:
```
Please type the whole signature when adding type annotations.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+
+        self._available = False
+        self._state = None
+        self._state_attrs = {ATTR_MODEL: self._model}
```
We don't add static info in the state attributes. We have device info where we set model.

jbouwh replied:
{ATTR_MODEL: self._model} was not needed and removed here.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+            self._humidity_steps = 10
+
+        self._state_attrs.update(
+            {attribute: None for attribute in self._available_attributes}
```
We don't set the item at all in the state attributes if the values is None.

jbouwh replied:
Items are actual initiated with None value.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+        """Return the target humidity."""
+        if self._state:
+            return (
+                self._state_attrs[ATTR_TARGET_HUMIDITY]
```
We shouldn't duplicate information in the extra state attributes that we also set in state properties of entities. The base entity of the platform integration will normally set those as state attributes anyway.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,382 @@
+    async_add_entities(entities, update_before_add=True)
+
+
+class XiaomiGenericHumidifierDevice(XiaomiMiioEntity, HumidifierEntity):
```
Please name this something Entity or just remove Device at the end.

jbouwh replied:
Is now XiaomiGenericHumidifier

jbouwh commented on homeassistant/components/xiaomi_miio/device.py:
```diff
@@ -53,17 +55,20 @@ async def async_connect_device(self, host, token):
-    def __init__(self, name, device, entry, unique_id):
+    def __init__(self, name, device, entry, unique_id, coordinator=None):
         """Initialize the Xiaomi Miio Device."""
+        if coordinator:
```
I agree, but I suggest to make this conditional to prevent this PR is becoming to big. Another PR https://github.com/home-assistant/core/pull/51791 that updates the fan platform and also uses this entity class `XiaomiMiioEntity`. For this integration should be merged fist. Further the class `XiaomiMiioEntity` is used in the platforms `air_quality`, `light`, `switch` and `vacuum`.
This PR covers using the coordinator for `humidifier` and the new code in `sensor`. I suggest to leave that for now. I could make an inline remark in the code that the use of the coordinator should be mandantory for all platform setups.

MartinHjelmare replied:
Just create a new class for the coordinator entity that also inherits the old base class. We should not make the coordinator optional here.

jbouwh replied:
Created an extra specialized class `XiaomiCoordinatedMiioEntity`

jbouwh committed:
Not await create_miio_device_and_coordinator

jbouwh committed:
Not await create_miio_device_and_coordinator

jbouwh commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +83,51 @@ def get_platforms(config_entry):
     return []
 
 
+async def async_create_miio_device_and_coordinator(
```
Done, also updated async_get_platforms() to get_platforms, since no IO is happening there neither.

jbouwh committed:
Invert model check

jbouwh committed:
Invert model check

jbouwh commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +83,51 @@ def get_platforms(config_entry):
+    token = entry.data[CONF_TOKEN]
+    name = entry.title
+    device = None
+    if model in MODELS_HUMIDIFIER:
```
Done

jbouwh committed:
use local device reference

jbouwh committed:
use local device reference

jbouwh commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +83,51 @@ def get_platforms(config_entry):
+
+    async def async_update_data():
+        """Fetch data from the device using async_add_executor_job."""
+        device = hass.data[DOMAIN][entry.entry_id][KEY_DEVICE]
```
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -95,40 +95,40 @@ def create_miio_device_and_coordinator(
-    if model in MODELS_HUMIDIFIER:
+    if model not in MODELS_HUMIDIFIER:
+        return
+    else:
```
Remove the else and outdent below.

jbouwh committed:
specialize XiaomiMiioEntity for coordinator+pylint

jbouwh committed:
specialize XiaomiMiioEntity for coordinator+pylint

jbouwh committed:
return a value specifically when returning

jbouwh committed:
return a value specifically when returning

jbouwh committed:
Rename XiaomiGenericHumidifierDevice

jbouwh committed:
Rename XiaomiGenericHumidifierDevice

jbouwh committed:
type the whole signature

jbouwh committed:
type the whole signature

jbouwh committed:
remove ATTR_MODEL - fix refresh

jbouwh committed:
remove ATTR_MODEL - fix refresh

jbouwh committed:
rename or remove _state_attrs

jbouwh committed:
rename or remove _state_attrs

jbouwh committed:
Trigger tests

starkillerOG commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +85,50 @@ def get_platforms(config_entry):
+def create_miio_device_and_coordinator(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+):
+    """Set up a data coordinator and one mio device to service multiple entities."""
```
mio --> miio

jbouwh replied:
done

starkillerOG commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "Xiaomi Miio Device"
+DATA_KEY = "fan.xiaomi_miio"
```
I don't think we need DATA_KEY

jbouwh replied:
Clean it up and all referencing code

starkillerOG commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "Xiaomi Miio Device"
```
DEFAULT_NAME is not used anymore --> remove

jbouwh replied:
done

starkillerOG commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+DEFAULT_NAME = "Xiaomi Miio Device"
+DATA_KEY = "fan.xiaomi_miio"
+
+CONF_MODEL = "model"
```
import from .const

jbouwh replied:
Done

starkillerOG commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+    if not config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+        return
+
+    if DATA_KEY not in hass.data:
```
DATA_KEY Not needed

jbouwh replied:
Removed

starkillerOG commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+            hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
+        )
+
+    hass.data[DATA_KEY][host] = entity
```
We do not need to store the entity.
Remove this DATA_KEY

jbouwh replied:
Done

starkillerOG commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -59,21 +70,23 @@
+ATTR_HUMIDITY = "humidity"
 
-SUCCESS = ["ok"]
+CONF_MODEL = "model"
```
import from .const

jbouwh replied:
done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/device.py:
```diff
@@ -89,3 +92,35 @@ def device_info(self):
         return device_info
+
+
+class XiaomiCoordinatedMiioEntity(CoordinatorEntity, XiaomiMiioEntity):
```
Let's not inherit from `XiaomiMiioEntity` for now and copy its properties to `XiaomiCoordinatedMiioEntity`. We can clean up and refactor later.

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Humidifier from a config entry."""
+    entities = []
+
+    if not config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+        return
```
```suggestion
    if not config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
        return

    entities = []
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+    if not config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+        return
+
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
+
```
```suggestion
```

jbouwh replied:
Removed

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+            hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
+        )
+
+    hass.data[DATA_KEY][host] = entity
```
```suggestion
```

jbouwh replied:
Was removed

starkillerOG reviewed (commented):
Looks very good to me :)

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +85,50 @@ def get_platforms(config_entry):
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(seconds=10),
+    )
+
```
Should we do the first data fetch (`coordinator.async_refresh`) here already, instead of in the platforms?

jbouwh replied:
async_request_refresh is only needed to trigger an update after a change was made to the device, Every 10 sec the device is polled and updated in the platforms already. I have added an initial data fetch now, it requires to await the function because triggering should be awaited.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+    @property
+    def available(self):
+        """Return true when state is known."""
+        return self._available
```
Check the `available` property of the parent coordinator entity too.

jbouwh replied:
done, checking and self.coordinator.last_update_success. The parent coordinator has no available property

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+        self._skip_update = False
+        self._available_modes = []
+        self._mode = None
+        self._available_attributes = AVAILABLE_ATTRIBUTES
```
We can use the constant directly. We don't need the instance attribute.

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +175,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        self.coordinator.async_add_listener(self.update)
+
+    @callback
+    def update(self):
```
Overwrite this method instead:
https://github.com/home-assistant/core/blob/c149ecf2cc6a4efef00d537baed200b9f96bd4ed/homeassistant/helpers/update_coordinator.py#L322-L324
```suggestion
    def _handle_coordinator_update(self):
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +175,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        _LOGGER.debug("Got new state: %s", state)
+        self._available = True
+        self._state = state.is_on
+        self._state = self._extract_value_from_attribute(state, self._attribute)
```
```suggestion
        self._state = self._extract_value_from_attribute(state, self._attribute)
        self.async_write_ha_state()
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +175,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        self._available = None
+        self._state = None
+        self._skip_update = False
+        self.coordinator.async_add_listener(self.update)
```
```suggestion
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+                key: self._extract_value_from_attribute(state, value)
+                for key, value in self._available_attributes.items()
+            }
+        )
```
```suggestion
        )
        self.async_write_ha_state()
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+    token = config_entry.data[CONF_TOKEN]
+    name = config_entry.title
+    model = config_entry.data[CONF_MODEL]
+    unique_id = config_entry.unique_id
```
```suggestion
    unique_id = config_entry.unique_id
    coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+            air_humidifier,
+            config_entry,
+            unique_id,
+            hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
```
```suggestion
            coordinator,
```

jbouwh replied:
Done

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,387 @@
+            air_humidifier,
+            config_entry,
+            unique_id,
+            hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
```
```suggestion
            coordinator,
```

jbouwh replied:
Done

MartinHjelmare commented:
You can batch commit the suggestions directly here in the GitHub UI from the files changed tab.

jbouwh committed:
Updates suggested by starkillerOG MartinHjelmare

jbouwh committed:
Updates suggested by starkillerOG MartinHjelmare

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -115,15 +107,14 @@ def __init__(self, name, device, entry, unique_id, coordinator):
     def available(self):
         """Return true when state is known."""
-        return self._available
+        return self._available and self.coordinator.last_update_success
```
```suggestion
        return super().available and self._available
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -265,7 +264,7 @@ def state_class(self):
     def available(self):
         """Return true when state is known."""
-        return self._available
+        return self._available and self.coordinator.last_update_success
```
```suggestion
        return super().available and self._available
```

jbouwh committed:
Update homeassistant/components/xiaomi_miio/device.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/device.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/humidifier.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/humidifier.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bieniu commented:
`MODELS_ALL_DEVICES` const should be changed, currently config flow argues that all humidifiers are unsupported.
```python
MODELS_ALL_DEVICES = (
    MODELS_SWITCH
    + MODELS_VACUUM
    + MODELS_AIR_MONITOR
    + MODELS_FAN
    + MODELS_LIGHT
    + MODELS_HUMIDIFIER
)
```

MartinHjelmare commented:
The breaking changes and PR description are not completely clear to me. What do we mean with this?

> This PR does not touch Xiaomi's fan platform, so this enables users to change their configuration. They can simple disable not used entities.

jbouwh committed:
Add MODELS_HUMIDIFIER to MODELS_ALL_DEVICES

jbouwh committed:
Add MODELS_HUMIDIFIER to MODELS_ALL_DEVICES

jbouwh commented:
> 
> 
> The breaking changes and PR description are not completely clear to me. What do we mean with this?
> 
> > This PR does not touch Xiaomi's fan platform, so this enables users to change their configuration. They can simple disable not used entities.

The related PR  https://github.com/home-assistant/core/pull/51791 updates the fan platform. The Airhumidiers still need to be be removed from fan.py. The users arirhumidifiers humidifier fan entities will not be available and should be removed.

bieniu commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +85,50 @@ def get_platforms(config_entry):
+    device = None
+    if model not in MODELS_HUMIDIFIER:
+        return
+    if model in MODELS_HUMIDIFIER_MIOT:
+        device = AirHumidifierMiot(host, token)
+    else:
+        device = AirHumidifier(host, token, model=model)
```
I suggest removing the fan entity for humidifiers.

```suggestion
    if model in MODELS_HUMIDIFIER_MIOT:
        device = AirHumidifierMiot(host, token)
    else:
        device = AirHumidifier(host, token, model=model)
        
    # removing fan platform entity for humidifiers
    entity_registry = await er.async_get_registry(hass)
    entity_id = entity_registry.async_get_entity_id("fan", DOMAIN, entry.unique_id)
    if entity_id:
        entity_registry.async_remove(entity_id)
```

jbouwh replied:
`They can simple disable not used entities.` removed from PR description. Entity registry import added.

jbouwh committed:
Update homeassistant/components/xiaomi_miio/__init__.py

Removing the fan entity for humidifiers.

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/__init__.py

Removing the fan entity for humidifiers.

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

bieniu commented:
> The Airhumidiers still need to be be removed from fan.py

IMO humidifiers should be removed from `fan.py` in this PR.

jbouwh committed:
Import entity registry

jbouwh committed:
Import entity registry

jbouwh commented:
> 
> 
> > The Airhumidiers still need to be be removed from fan.py
> 
> IMO humidifiers should be removed from `fan.py` in this PR.

I will have a look at this tonight

jbouwh committed:
Cleanup fan platform code

jbouwh committed:
Cleanup fan platform code

jbouwh commented:
I cleaned up the code for the fan platform. Since related PR #51791 improves the fan platform code I made no further changes so that merging PR #51791 will be easier.

jbouwh commented:
> 
> 
> The breaking changes and PR description are not completely clear to me. What do we mean with this?
> 
> > This PR does not touch Xiaomi's fan platform, so this enables users to change their configuration. They can simple disable not used entities.

Since the fan code is cleaned up as well this line was removed from the description.

bieniu commented:
I get an exception when I try to change the mode:
```
Traceback (most recent call last):
  File "/home/maciek/develop/home-assistant-core/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/home/maciek/develop/home-assistant-core/homeassistant/core.py", line 1491, in async_call
    task.result()
  File "/home/maciek/develop/home-assistant-core/homeassistant/core.py", line 1526, in _execute_service
    await handler.job.target(service_call)
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity_component.py", line 213, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/service.py", line 658, in entity_service_call
    future.result()  # pop exception if have
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity.py", line 811, in async_request_call
    await coro
  File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/service.py", line 695, in _handle_entity_call
    await result
  File "/home/maciek/develop/home-assistant-core/homeassistant/components/xiaomi_miio/humidifier.py", line 293, in async_set_mode
    if mode not in AirhumidifierOperationMode:
  File "/usr/local/lib/python3.9/enum.py", line 373, in __contains__
    raise TypeError(
TypeError: unsupported operand type(s) for 'in': 'str' and 'EnumMeta'
```

I thing line 291 should looks like this:
```python
if mode not in AirhumidifierOperationMode.__members__.keys():
```

jbouwh committed:
Fix set mode for XiaomiAirHumidifier

jbouwh committed:
Fix set mode for XiaomiAirHumidifier

jbouwh commented:
I removed the classes 

> 
> 
> I get an exception when I try to change the mode:
> 
> ```
> Traceback (most recent call last):
>   File "/home/maciek/develop/home-assistant-core/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
>     await hass.services.async_call(
>   File "/home/maciek/develop/home-assistant-core/homeassistant/core.py", line 1491, in async_call
>     task.result()
>   File "/home/maciek/develop/home-assistant-core/homeassistant/core.py", line 1526, in _execute_service
>     await handler.job.target(service_call)
>   File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity_component.py", line 213, in handle_service
>     await self.hass.helpers.service.entity_service_call(
>   File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/service.py", line 658, in entity_service_call
>     future.result()  # pop exception if have
>   File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/entity.py", line 811, in async_request_call
>     await coro
>   File "/home/maciek/develop/home-assistant-core/homeassistant/helpers/service.py", line 695, in _handle_entity_call
>     await result
>   File "/home/maciek/develop/home-assistant-core/homeassistant/components/xiaomi_miio/humidifier.py", line 291, in async_set_mode
>     if mode not in AirhumidifierOperationMode:
>   File "/usr/local/lib/python3.9/enum.py", line 373, in __contains__
>     raise TypeError(
> TypeError: unsupported operand type(s) for 'in': 'str' and 'EnumMeta'
> ```
> 
> I thing line 291 should looks like this:
> 
> ```python
> if mode not in AirhumidifierOperationMode.__members__.keys():
> ```

I was able to reproduce the error and committed a fix. I could only fysically test XiaomiAirHumidifierMiot.

jbouwh commented:
For XiaomiAirHumidifierMiot mode Auto is actually using `target_humidity` but for XiaomiAirHumidifier V1 there is no mode Auto. When changing `target_humidity` for XiaomiAirHumidifierMiot the code will set mode to `auto`. In other modes the target_humidity will be not be shown. This works for XiaomiAirHumidifierMiot . What should be the correct behavior for XiaomiAirHumidifier , especialliy for XiaomiAirHumidifier V1, that has not an `Auto` mode?

jbouwh committed:
Do not set auto mode if not supported

jbouwh committed:
Do not set auto mode if not supported

rytilahti commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -95,6 +107,23 @@ class SensorType:
+    ),
+}
+
+HUMIDIFIER_ATTRIBUTES = {
```
```suggestion
HUMIDIFIER_SENSORS = {
```
maybe?

jbouwh replied:
Updated

rytilahti commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +174,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+            attributes = HUMIDIFIER_ATTRIBUTES
+        if device:
+            for attribute in attributes:
+                entities.append(
```
This structure doesn't look very nice. A nicer way would be to have a "device" always available that contains either an empty list of sensors ("attributes") and handle this in a generic manner.

There will be much more sensors exposed by other platforms in the future, so it would be great to get this done properly to make it simple to convert other platforms away from state attributes.

jbouwh replied:
I do not want risking with touching the AirQualityMonitor to much in this PR. This could be updated later too. For now I have made adding sensor generic and took it out of the if then construct.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +174,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+            return
+        _LOGGER.debug("Got new state: %s", state)
+        self._available = True
+        self._state = state.is_on
```
Can we remove this line?
```suggestion
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,21 +174,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        device = None
+        if model in MODELS_HUMIDIFIER_MIOT:
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
```
The coordinator is the same for all these cases.

jbouwh replied:
Yes, but `hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]` is not initialized for the AirQualityMonitor.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,378 @@
+ATTR_TRANS_LEVEL = "trans_level"
+ATTR_HARDWARE_VERSION = "hardware_version"
+
+SERVICE_TO_METHOD = {
```
This isn't used.

jbouwh replied:
Removed unused constants.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/humidifier.py:
```diff
@@ -0,0 +1,378 @@
+
+        _LOGGER.debug("Setting the humidity to: %s", target_humidity)
+        await self._try_command(
+            "Setting operation mode of the miio device failed.",
```
Shouldn't this be "Setting the target humidity"?

jbouwh replied:
updated

MartinHjelmare commented on homeassistant/components/xiaomi_miio/fan.py:
```diff
@@ -1037,196 +952,6 @@ async def async_set_led_brightness(self, brightness: int = 2):
-            humidity,
-        )
-
-    async def async_set_dry_on(self):
```
How do we support dry mode, led brightness and motor speed services now?

bieniu replied:
The question is whether we should use services or maybe entities (switch for dry mode, select for led brightness, etc.). Motor speed needs a service for sure.

jbouwh replied:
I think we using entities to switch or control will be better, we could add those later. _Let try to make this PR to big._ For now I have only removed the humidifier code from fan.py to make merging of the fan related PR easier.
We could add the additional entities for motor_speed, led_brightness, dry_mode, buzzer and child_lock later. Finally we should remove _Humidifier only_ services since they cannot be used any more. I will remove the AIRHUMIDIFIER related feature flags and FEATURES for now. What services need to be removed that are AIRHUMIDIFIER  specific?

bieniu replied:
Yes, I totally agree that adding another platforms should be in the next PRs.
I think those services are for humidifiers only and should be removed from `fan.py`: `SERVICE_SET_TARGET_HUMIDITY`, `SERVICE_SET_DRY_ON`, `SERVICE_SET_DRY_OFF`, `SERVICE_SET_MOTOR_SPEED`

jbouwh replied:
`dry_mode`, `child_lock` and `buzzer` could later be implemented as `switch` https://www.home-assistant.io/integrations/switch/. Implementing `switch.turn_on` and `switch.turn_off` will remove the need for specific services.
Motor control could be implemented using `input_number` https://www.home-assistant.io/integrations/input_number/
This will also allow using a slider.
The led brightness could be implemented using an input select: https://www.home-assistant.io/integrations/input_select/.

bieniu replied:
I think `input_select` and `input_number` can't be used by integrations, but maybe I'm wrong.
We have now `select` platform https://github.com/home-assistant/core/pull/51849

jbouwh replied:
Removed not supported services.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/fan.py:
```diff
@@ -1037,196 +952,6 @@ async def async_set_led_brightness(self, brightness: int = 2):
-        return None
-
-    @property
-    def button_pressed(self):
```
Can we expose this as a sensor instead?

jbouwh replied:
We could, but what does it add? It does not seem usefull to me.

jbouwh replied:
To demonstrate how a `button_pressed` sensor would look see the screenshot.
![afbeelding](https://user-images.githubusercontent.com/7188918/122643913-b6b71380-d112-11eb-9b95-687605f8509c.png)
The `button_pressed` attribute only indicates the last button pressed. If the same button is pressed again, there will be not state change. I don't think this is very useful. Not all humidifier models support the attribute, so that would need additional logic as well. But as you can see adding sensor is easy now.

bieniu commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +85,56 @@ def get_platforms(config_entry):
+
+    # removing fan platform entity for humidifiers
+    entity_registry = await en.async_get_registry(hass)
+    entity_id = entity_registry.async_get_entity_id("fan", DOMAIN, entry.unique_id)
```
These methods aren't async but are marked safe to be called from the event loop:
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity_registry.py#L189
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity_registry.py#L319

jbouwh replied:
Added await for async calls not returning values., object str can't be used in `await` expression.

jbouwh commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -75,6 +85,56 @@ def get_platforms(config_entry):
+    token = entry.data[CONF_TOKEN]
+    name = entry.title
+    device = None
+    if model not in MODELS_HUMIDIFIER:
+        return
```
Something like:
```python
_LOGGER.error(
        "Unsupported device found! Please create an issue at "
        "https://github.com/syssi/xiaomi_airpurifier/issues "
        "and provide the following data: %s",
        model,
)
```

Is in e.g. fan.py. We should not put this in __init__.py for now because then we should check all device models, we log an error in de platform code instead. The URL for issue e.g. might be some confusing.
Since async_create_miio_device_and_coordinator is always called as a generic device setup it should log and an error here, since that would also impact other device classes. There where we want to use a coordinator we can add additional models here. If we want to centralize this we could use `get_platforms()`, and log an error there there, but we should do remove the logging in the platform code.

jbouwh replied:
I have added a line that logs an error if no platform was found to be set up. I think that should cover it.

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Update docstring for SensorType class

Co-authored-by: Teemu R. <tpr@iki.fi>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Update docstring for SensorType class

Co-authored-by: Teemu R. <tpr@iki.fi>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Deleted ineffective code

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Deleted ineffective code

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Minor corrections

jbouwh committed:
Minor corrections

jbouwh committed:
Merge branch 'xiaomi-miio-humidifier-entity' of https://github.com/jbouwh/core into xiaomi-miio-humidifier-entity

bieniu commented:
I'm wondering, according to [this](https://github.com/rytilahti/python-miio/blob/master/miio/airhumidifier_miot.py#L131-L132), should we return `water_level` sensor state only when `value <= 100`? A value greater than `100` means the device cover is removed.

jbouwh commented:
> 
> 
> I'm wondering, according to [this](https://github.com/rytilahti/python-miio/blob/master/miio/airhumidifier_miot.py#L131-L132), should we return `water_level` sensor state only when `value <= 100`? A value greater than `100` means the device cover is removed.

Then it will return e.d. 120% the way it implemented now. Would that be a problem?

bieniu commented:
> Would that be a problem?

Problem, no, but 120% as a state of `water level` sensor is misleading in my opinion. We can fix this in the future I think, maybe sensor should be `unavailable` when `value > 100`.

bieniu commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -159,7 +164,7 @@ async def async_setup_gateway_entry(
 
     device_registry = await dr.async_get_registry(hass)
-    device_registry.async_get_or_create(
+    await device_registry.async_get_or_create(
```
This method is not async https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/device_registry.py#L243

jbouwh replied:
Ill revert that

jbouwh committed:
Report unknown value is sensor is out of range

jbouwh committed:
Report unknown value is sensor is out of range

jbouwh commented:
> 
> 
> > Would that be a problem?
> 
> Problem, no, but 120% as a state of `water level` sensor is misleading in my opinion. We can fix this in the future I think, maybe sensor should be `unavailable` when `v

> 
> 
> > Would that be a problem?
> 
> Problem, no, but 120% as a state of `water level` sensor is misleading in my opinion. We can fix this in the future I think, maybe sensor should be `unavailable` when `value > 100`.

See latest commit. Made it generic.

jbouwh committed:
Avoid assigning the default values

jbouwh committed:
Avoid assigning the default values

jbouwh committed:
Remove not supported fan services

jbouwh committed:
Remove not supported fan services

jbouwh committed:
Reduce device polling rate

jbouwh committed:
Reduce device polling rate

jbouwh commented:
If have set the device polling rate of the data coordinator at 60 sec. After commands are set, de device data is refreshed any way.

bieniu commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -199,7 +199,7 @@ async def async_update_data():
         update_method=async_update_data,
         # Polling interval. Will only be polled if there are subscribers.
-        update_interval=timedelta(seconds=10),
+        update_interval=timedelta(seconds=60),
```
This is the update interval for the gateway coordinator.

jbouwh replied:
I know. Does it really need need to poll every 10 seconds?

bieniu replied:
Honestly, I'm not sure. In this PR we dealing with miio humidifier so we shoudn't touch gateway code. You can change update interval for humidifiers [here](https://github.com/home-assistant/core/blob/496ba8f77fa85b85ef5475ee408c3019f58aac12/homeassistant/components/xiaomi_miio/__init__.py#L133).

jbouwh replied:
You are right, my bad.

jbouwh replied:
My intention was to update the polling interval for devices. It will update every 60 is no commands are given. It will also be called when a refresh is demanded. So 60 sec. should be more then enough to update the sensors.

jbouwh committed:
Rate interval for device, not gateway

jbouwh committed:
Rate interval for device, not gateway

bieniu commented:
> See latest commit. Made it generic.

I have another idea how to handle this. We can use a new class for `water_level` sensor:
```python
class XiaomiWaterLevelSensor(XiaomiGenericSensor):
    """Representation of a Xiaomi Humidifier Water Level sensor."""
    def __init__(self, name, device, entry, unique_id, attribute, coordinator):
        """Initialize the entity."""
        super().__init__(name, device, entry, unique_id, attribute, coordinator)

    @property
    def available(self):
        """Return true when state is known."""
        return super().available and self._state <= 100
```
So, if the sensor state is greater than 100, the sensor will be `unavailable`.

Of course add entities logic should be changed:
```python
        for sensor in sensors:
            if sensor == ATTR_WATER_LEVEL:
                entities.append(
                    XiaomiWaterLevelSensor(
                        f"{config_entry.title}_{sensor}",
                        device,
                        config_entry,
                        f"{sensor}_{config_entry.unique_id}",
                        sensor,
                        hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
                    )
                )
            else:
                entities.append(
                    XiaomiGenericSensor(
                        f"{config_entry.title}_{sensor}",
                        device,
                        config_entry,
                        f"{sensor}_{config_entry.unique_id}",
                        sensor,
                        hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
                    )
                )
```
and this code in `state` property of `XiaomiGenericSensor` class will not be needed:
```python
        if (
            SENSOR_TYPES[self._attribute].valid_min_value
            and self._state < SENSOR_TYPES[self._attribute].valid_min_value
        ):
            return None
        if (
            SENSOR_TYPES[self._attribute].valid_max_value
            and self._state > SENSOR_TYPES[self._attribute].valid_max_value
        ):
            return None
```

What do you think about this?

jbouwh commented:
I would prefer the value range check with a generic sensor class, since the class can be reused. Specializing the water_level sensor does not add much to my opinion.

I could make the sensor `not` available not changing the state property code, but I would state that `Unknown` is a better representation then `not available` when the cover is open. Additional a binairy sensor for cover open could indicate the cover is open.

Th only advantage of a specialized class would be that it does not heed to check for minimum or maximum values at each state fetch.

jbouwh commented:
My advice for now would be to finish this PR first now so we can finish the work on the fan speeds, further we can add extra controls for motor_speed e.g.

bieniu commented on homeassistant/components/xiaomi_miio/device.py:
```diff
@@ -89,3 +92,60 @@ def device_info(self):
+        device_info = {
+            "identifiers": {(DOMAIN, self._device_id)},
+            "manufacturer": "Xiaomi",
+            "name": self._name,
```
```suggestion
            "name": self._device_name,
```

![device_name](https://user-images.githubusercontent.com/478555/122745458-50a0cc80-d289-11eb-9b0b-fc7df7af9e2d.png)

bieniu commented on homeassistant/components/xiaomi_miio/device.py:
```diff
@@ -89,3 +92,60 @@ def device_info(self):
+        self._device = device
+        self._model = entry.data[CONF_MODEL]
+        self._mac = entry.data[CONF_MAC]
+        self._device_id = entry.unique_id
```
```suggestion
        self._device_id = entry.unique_id
        self._device_name = entry.title
```

bieniu commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -145,19 +175,119 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+        for sensor in sensors:
+            entities.append(
+                XiaomiGenericSensor(
+                    f"{config_entry.title}_{sensor}",
```
```suggestion
                    f"{config_entry.title} {sensor.title()}",
```

bieniu replied:
Ups, sorry for the quote mistake.

bieniu reviewed (commented):
Few small comments.

jbouwh committed:
Update homeassistant/components/xiaomi_miio/__init__.py

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/__init__.py

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/sensor.py

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/device.py

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Update homeassistant/components/xiaomi_miio/device.py

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

jbouwh committed:
Device name in XiaomiCoordinatedMiioEntity

jbouwh committed:
Device name in XiaomiCoordinatedMiioEntity

jbouwh committed:
Quote syntax error

jbouwh committed:
Quote syntax error

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -71,10 +81,65 @@ def get_platforms(config_entry):
+        KEY_COORDINATOR: coordinator,
+    }
+    # Trigger first data fetch
+    await coordinator.async_config_entry_first_refresh()
```
This will fail the whole integration setup and retry it later. Is that what we want even if it's just one device that isn't connected ok?

bieniu replied:
But the coordinator is created for config entry, that is for one device. Am I missing something?

jbouwh replied:
The setup is repeated for every device in the network. It should fail if no data could be fetched.

MartinHjelmare commented:
Code wise this looks good to me. I'm a bit worried about the removal of the custom services where we don't have a replacement yet. I don't think we should release this until we have a replacement for those. The beta cut is in 1.5 weeks. What do you think?

bieniu reviewed (approved):
LGTM, great work 👍

bieniu commented:
> What do you think?

Maybe temporarily merge this change to a separate branch and make it available to users only after all necessary platforms are added? If we don't make it before the beta cut, we'll wait until 2021.8.

jbouwh commented:
> > What do you think?
> 
> Maybe temporarily merge this change to a separate branch and make it available to users only after all necessary platforms are added? If we don't make it before the beta cut, we'll wait until 2021.8.

The extra services for the humidifier do not really add much value, they should not be a show stopper. I would merge into dev, so we can go on fixing fan speeds that are te deprecated.

MartinHjelmare commented:
If users are using these services I assume they would disagree with that. This is one of the most used integrations according to our analytics so we should be careful with breaking things.

jbouwh commented:
> 
> 
> If users are using these services I assume they would disagree with that. This is one of the most used integrations according to our analytics so we should be careful with breaking things.

True, but child_lock, buzzer, dry_mode services we use easily implement using the switch integration.
For motor_speed and led_brightness I am already working on code in the same line as the sensors were implemented.

jbouwh commented:
Okay. I already [have some new code committed](https://github.com/jbouwh/core/tree/xiamoi-miio-humidifier-controls) for the select platform, supporting Led Brightness for humidifiers. It uses The [`EntitySelect`](https://developers.home-assistant.io/docs/core/entity/select/ ) class that was recently merged into `dev`. So we need to rebase with `dev` anyway to be able add the new code. Further for PR #51791 it should be considered it is to be merged into dev first, or that we add it to this new branch as well. Again the code will support the missing humidifier services first. The other devices will keep their legacy services until the humidifier services have been replaced. I can do physically testing with the `MODEL_AIRHUMIDIFIER_CA4` which I have setup here.

bieniu commented:
I tested `select` platform with my `zhimi.humidifier.ca1`. It works but `select` entity state change is very slow. I think `await self.coordinator.async_request_refresh()` is needed after `try_command()`.

jbouwh commented:
> I tested `select` platform with my `zhimi.humidifier.ca1`. It works but `select` entity state change is very slow. I think `await self.coordinator.async_request_refresh()` is needed after `try_command()`.

Try again, I fixed a sync problem. Further I implemented the switch platform now.

jbouwh commented:
In the mean while I implemented motor_speed (using `NumberEntity`) and an additional sensor `actual_apeed` as well at https://github.com/jbouwh/core/tree/xiamoi-miio-humidifier-controls. I think we would have all useful controls back if this was merged.

MartinHjelmare commented:
I've pushed a fresh copy of dev branch here:
https://github.com/home-assistant/core/tree/rework-xiaomi-miio-fans-humidifiers

Please rebase on top of that branch here and then change the target branch to that branch in this PR.

We can than make multiple PRs to that branch and when happy with the new fan and humidifier support @jbouwh can make a PR from that branch to dev branch.

jbouwh committed:
Merge branch 'xiaomi-miio-humidifier-entity' of https://github.com/jbouwh/core into xiaomi-miio-humidifier-entity

MartinHjelmare commented:
There are some unused constants that remain, eg `SERVICE_SCHEMA_TARGET_HUMIDITY`. Please go over the constants and remove the ones that are no longer used.

Side note: As a future follow up it would be good to go over all the extra state attributes in the fan platform and move the ones that don't belong there to separate sensors etc, eg humidity, temperature, air quality index etc.

jbouwh committed:
Clean up constants

jbouwh commented:
> 
> 
> There are some unused constants that remain, eg `SERVICE_SCHEMA_TARGET_HUMIDITY`, at least in the fan platform. Please go over the constants and remove the ones that are no longer used.
> 
> Side note: As a future follow up it would be good to go over all the extra state attributes in the fan platform and move the ones that don't belong there to separate sensors etc, eg humidity, temperature, air quality index etc.

Done. More constants will be moved to const.py later because the will be reused for more platforms.

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Please confirm that we can merge here now.

jbouwh commented:
> 
> 
> Please confirm that we can merge here now.

We are good to go!

MartinHjelmare commented:
Side note: We'll need to update the docs before merging the rework to dev.
